### ISSUE TITLE
Reducer adding multiple lines of "precision highp float" when reducing a shader

### ISSUE THREAD

**Issue by Vihanakangas** [google/graphicsfuzz#927]
As shown in the example below, the reducer does not check if `precision highp float;` already exists in the variant shader, instead, it is added on every donor injection.

``` glsl
#version 310 es
precision highp float;

precision highp float;

precision highp float;

precision highp float;

precision highp float;

precision highp float;

precision highp float;

precision highp float;

layout(location = 0) out vec4 _GLF_color;

layout(set = 0, binding = 0) uniform buf0 {
 vec2 resolution;
};
const vec4 pal[16] = vec4[16](vec4(0.0, 0.0, 0.0, 1.0), vec4(0.5, 0.0, 0.0, 1.0), vec4(0.0, 0.5, 0.0, 1.0), (vec4(mat2(vec4(0.5, 0.5, 0.0, 1.0)))), vec4(0.0, 0.0, 0.5, 1.0), vec4(0.5, 0.0, 0.5, 1.0), vec4(0.0, 0.5, 0.5, 1.0), vec4(0.5, 0.5, 0.5, 1.0), vec4(0.0, 0.0, 0.0, 1.0), vec4(1.0, 0.0, 0.0, 1.0), vec4(0.0, 1.0, 0.0, 1.0), vec4(1.0, 1.0, 0.0, 1.0), vec4(0.0, 0.0, 1.0, 1.0), vec4(1.0, 0.0, 1.0, 1.0), vec4(0.0, 1.0, 1.0, 1.0), vec4(1.0, 1.0, 1.0, 1.0));

ivec2 iter(ivec2 p)
{
 if(p.x > 0)
  {
   p.y --;
  }
 if(p.x < 0)
  {
   p.y ++;
  }
 p.x += p.y / 2;
 return p;
}
void main()
{
 vec2 pos = gl_FragCoord.xy / resolution;
 ivec2 ipos = ivec2(int(pos.x * 8.0), int(pos.y * 8.0));
 int v = (ipos.x & 5) | (ipos.y & 10);
 int w = (ipos.y & 5) | (ipos.x & 10);
 ivec2 p = ivec2(v * 8 + w, 0);
 int i;
 for(
     i = 0;
     i < 100;
     i ++
 )
  {
   p = iter(p);
  }
 if(p.x < 0)
  {
   p.x = - p.x;
  }
 while(p.x > 15)
  {
   p.x -= 16;
  }
 _GLF_color = pal[p.x];
}
```

**Comment by Vihanakangas** [comment]
Same thing seems to be happening with `precision highp int;`-line as well. Probably caused by the same bug.

**Comment by afd** [comment]
When preserving semantics, we should allow (*) to be removed in the following scenario:

```
precision _ T; (*)
// other precision declarations
precision _ T; (**)
```

This is because whatever (*) specifies is over-ridden by (**).

When not preserving semantics, we can be more aggressive and remove either (*) or (**).  That's because if we change the default qualifiers due to removing one of these statements, the fact that this changes semantics does not matter.

What we must not do, even when changing semantics, is leave *no* precision qualifiers when some defaults are required.  In particular, in a fragment shader, we must not remove a float default precision qualifier if there is no other float default precision qualifier above it.

@jarisiru

**Comment by afd** [comment]
Fixed via #967.
